### PR TITLE
 Fix: libcrmcommon: Use truly implicit deny for ACLs

### PR DESCRIPTION
**Pull request by nrwahl2**
If a node does not have any ACLs set for it explicitly, it's assigned a
deny flag. However, in the case of scoped commands, the input to
`pcmk__apply_acl()` may be a section rather than the whole CIB. The
node at the root of the section that's passed in gets assigned a deny
flag after no ACLs match it.

This causes `pcmk__check_acl()` to fail prematurely when it walks up
the tree of parents. It reaches the node at the root of the section
that was passed to `pcmk__apply_acl()`, finds a deny flag, and returns
FALSE. If that node's parent (e.g., `/cib`) has a read or write ACL,
the parent's ACL is not reached and thus not considered.

The fix is to avoid setting a deny flag for a node if there is no ACL
for that node. `pcmk__check_acl()` already denies access implicitly if
neither read nor write is set for a node or any of its parents. The only
other change required is to have `xml_acl_filtered_copy()` check for the
absence of read and write instead of the presence of deny.

Resolves: RHBZ#1833173

### CODE REVIEW

File: lib/common/acl.c [hunk 1]
```diff
@@ -290,15 +290,6 @@ pcmk__apply_acl(xmlNode *xml)
     }
 
     p = xml->_private;
-    if (is_not_set(p->flags, xpf_acl_read)
-        && is_not_set(p->flags, xpf_acl_write)) {
-
-        p->flags |= xpf_acl_deny;
-        p = xml->doc->_private;
-        crm_info("Applied default deny ACL for user '%s' to <%s>",
```
Comment (nrwahl2) [inline lib/common/acl.c:298]: We may still want to keep this message or something like it. We're no longer **applying a default ACL**, but it's implicitly applied.

**Comment by kgaillot** [comment]
tests look good

File: lib/common/acl.c (hunk 1 above)
Comment (nrwahl2) [reply lib/common/acl.c:298]: I don't actually remember whether I considered that at the time or not.

This approach in pcmk__apply_acl seemed to mesh well with the way pcmk__check_acl is already set up. pcmk_check_acl walks up the parent tree looking for an explicit denial. If it doesn't find one, it enforces an implicit denial.

Walking up the parents requires little bit of additional code -- nothing too complicated, I think walk up until we find some ACL set or until we reach the root of the tree. Whereas we can achieve the same result with a bit less code (and a more natural feeling IMO) this way: applying an ACL only if the user configured one, and then denying by default later in processing if no ACL is set.

Comment (nrwahl2) [reply lib/common/acl.c:298]: As a side note, I just noticed that the old version logs "Applied default deny" even when there's an explicit deny ACL set.

File: lib/common/acl.c [hunk 2]
```diff
@@ -679,8 +670,9 @@ pcmk__check_acl(xmlNode *xml, const char *name, enum xml_private_flags mode)
                 return TRUE;
 
             } else if (is_set(p->flags, xpf_acl_deny)) {
-                crm_trace("Parent ACL denies user '%s' %s access to %s",
-                          docp->user, __xml_acl_to_text(mode), buffer);
+                crm_trace("%sACL denies user '%s' %s access to %s",
+                          (parent != xml) ? "Parent " : "", docp->user,
+                          __xml_acl_to_text(mode), buffer);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/acl.c:675]: Done, will push soon.

File: lib/common/acl.c [hunk 3]
```diff
@@ -489,8 +478,10 @@ xml_acl_filtered_copy(const char *user, xmlNode *acl_source, xmlNode *xml,
     }
 
     p = target->_private;
-    if (is_set(p->flags, xpf_acl_deny)
+    if (is_not_set(p->flags, xpf_acl_read)
+        && is_not_set(p->flags, xpf_acl_write)
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/acl.c]: Huh, it sure does. That's a useful function :) I'll remove those redundant checks.

**Comment by kgaillot** [comment]
Great work, thanks

**Pull request merged**